src/Orleans/Telemetry/MetricsWriter.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+using Orleans.Runtime.Configuration;
+
+namespace Orleans.Runtime
+{
+    public class MetricsWriter : IMetricsWriter, IDisposable
+    {
+        private List<IMetricTelemetryConsumer> consumers;
+
+        public MetricsWriter(IEnumerable<IMetricTelemetryConsumer> consumers)
+        {
+            this.consumers = consumers.ToList();
+        }
+
+        internal static MetricsWriter FromConfiguration(MetricsTelemetryConfiguration configuration, IServiceProvider serviceProvider)
+        {
+            var consumers = new List<IMetricTelemetryConsumer>(configuration.Consumers.Count);
+            foreach (var consumerConfig in configuration.Consumers)
+            {
+                IMetricTelemetryConsumer consumer = null;
+                if ((consumerConfig.Properties?.Count ?? 0) == 0)
+                {
+                    // first check whether it is registered in the container already
+                    consumer = (IMetricTelemetryConsumer)serviceProvider.GetService(consumerConfig.ConsumerType);
+                }
+                if (consumer == null)
+                {
+                    consumer = (IMetricTelemetryConsumer)ActivatorUtilities.CreateInstance(serviceProvider, consumerConfig.ConsumerType, consumerConfig.Properties?.Values?.ToArray() ?? new object[0]);
+                }
+                consumers.Add(consumer);
```

Comment:
Looks like using MetricTelemetryConfiguration in ClientConfig and NodeConfig is the suggested way to configure MetricsWriter. But using it means we need to use reflection to construct MetricTelemetryConsumer or ask application code to inject consumer type into DI. I think generally we should move away from using reflection and config.Properties to construct MetricTelemetryConsumer . For injecting consumer type into DI, it is not straightforward for user to figure this out, since neither MetricsWriter nor MetricTelemetryConfiguration has imply or enforced that programmatically.

Reply:
Why cannot we simply do 
  ```
  public class MetricsWriter : IMetricsWriter, IDisposable
    {
        public List<IMetricTelemetryConsumer> Consumers { get; set; }

        public MetricsWriter(IEnumerable<IMetricTelemetryConsumer> consumers)
        {
            this.Consumers = consumers.ToList();
        }
...remaining methods...
}
```

Making the configuration more programmatical? User can configure MetricWriter into DI using SiloBuilder

        private static IMetricsWriter CreateMetricWriter(IServiceProvider sp)
        {
            var consumer = new MetricConsumer(some param, sp.GetService<T2>);
            return new MetricsWriter(new List<IMetricTelemetryConsumer>{consumer});
        }

        public void ConfigureMetricWriter(IServiceCollection svc)
        {
            svc.AddSingleton<IMetricsWriter>(sp => CreateMetricWriter(sp));
            return svc;
        }

`siloBuilder.ConfigureServices(ConfigureMetricWriter)`

Reply:
Agree, we can improve it, but as mentioned in the description, this first pass is still back-compat WRT config. If I break compatibility in the same PR, we could hold it off for very long, as we need to make sure we have a way to bridge the gap with our existing users, as well as make sure the design is correct. And even with that, `master` still doesn't have a reference to `Microsoft.Extensions.Configuration`, so no point introducing it here either.

For all that, I don't think it's worth it to try to tackle it in the same PR if your plan is to rev the Logging part in the short term, since this configuration is 100% back compat for metric telemetry consumers (so I didn't introduce any new concepts, just made the previous config still work).

Reply:
sorry I missed that description. Mostly just want to get to know the plan more. I see MetricTelemetryConfiguration  is here to maintain backward compatibility in the transitioning phase. Will it be part of 2.0? Is it in 2.0. We support two ways of configuring MetricTelemetryConsumer, one is through CLusterconfig and Nodeconfig, another one is through SiloBuilder and ClientBuilder?

Reply:
Hopefully we can deprecate the ClusterConfiguration and ClientConfiguration object graphs soon. There is a small hint of what we could do with these god objects in https://github.com/dotnet/orleans/issues/3353 but there's no hard commitment on what to do with them for 2.0 (or if we will remove them after we ship 2.0).
That's why this PR is still separate from those decisions. For now the only way to configure telemetry is as it was before, through the ClusterConfiguration and ClientConfiguration. We can still enable the builder extension methods even if we don't deprecate these config objects yet (but again, separate PR, after logging settles).